radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -17,92 +17,235 @@
+	private static final Logger log = LogManager.getLogger();
+
+	public interface GetVerticesRequest {
+		Hash getVertexId();
+		int getCount();
+	}
+
+	public interface SyncSender {
+		void synced(Hash vertexId);
+	}
+
 	private final BehaviorSubject<Vertex> lastCommittedVertex = BehaviorSubject.create();
+	private final SyncSender syncSender;
+	private final SyncVerticesRPCSender syncVerticesRPCSender;
+	private final SyncedStateComputer<CommittedAtom> syncedStateComputer;
+	private final SystemCounters counters;
 
-	// Should never be null
-	private Vertex root;
+	// These should never be empty
+	private final AtomicReference<Hash> rootId = new AtomicReference<>();
+	private final AtomicReference<QuorumCertificate> highestQC = new AtomicReference<>();
+	private final AtomicReference<QuorumCertificate> highestCommittedQC = new AtomicReference<>();
 
-	// Should never be null
-	private volatile QuorumCertificate highestQC;
+	private final Map<Hash, Vertex> vertices = new ConcurrentHashMap<>();
+	private final Map<Hash, SyncState> syncing = new ConcurrentHashMap<>();
 
 	// TODO: Cleanup this interface
 	public VertexStore(
 		Vertex genesisVertex,
 		QuorumCertificate rootQC,
-		RadixEngine<LedgerAtom> engine,
+		SyncedStateComputer<CommittedAtom> syncedStateComputer,
+		SyncVerticesRPCSender syncVerticesRPCSender,
+		SyncSender syncSender,
 		SystemCounters counters
 	) {
-		this.engine = Objects.requireNonNull(engine);
+		this.syncedStateComputer = syncedStateComputer;
+		this.syncVerticesRPCSender = syncVerticesRPCSender;
+		this.syncSender = syncSender;
 		this.counters = Objects.requireNonNull(counters);
-		this.highestQC = Objects.requireNonNull(rootQC);
+		this.highestQC.set(Objects.requireNonNull(rootQC));
+		this.highestCommittedQC.set(rootQC);
 
 		if (genesisVertex.getAtom() != null) {
 			CommittedAtom committedGenesis = genesisVertex.getAtom().committed(rootQC.getProposed());
-			try {
-				this.engine.checkAndStore(committedGenesis);
-			} catch (RadixEngineException e) {
-				throw new IllegalStateException("Could not store genesis atom: " + genesisVertex.getAtom(), e);
-			}
+			syncedStateComputer.execute(committedGenesis);
 		}
+
+		this.rootId.set(genesisVertex.getId());
 		this.vertices.put(genesisVertex.getId(), genesisVertex);
-		this.root = genesisVertex;
 		this.lastCommittedVertex.onNext(genesisVertex);
 	}
 
-	public boolean syncToQC(QuorumCertificate qc) {
-		final Vertex vertex = vertices.get(qc.getProposed().getId());
-		if (vertex != null) {
+	private static class SyncState {
+		private final QuorumCertificate qc;
+		private final ECPublicKey author;
+		private final LinkedList<Vertex> fetched = new LinkedList<>();
+
+		SyncState(QuorumCertificate qc, ECPublicKey author) {
+			this.qc = qc;
+			this.author = author;
+		}
+	}
+
+	public void processGetVerticesRequest(GetVerticesRequest request) {
+		log.info("SYNC_VERTICES: Received GetVerticesRequest {}", request);
```

Comment:
This is definitely a place where we'll want to exercise care when we move into Byzantine land (consider nodes sending ridiculous requests, forcing us to gather and send a bunch of vertices). Maybe it should be marked as such for later consideration, what do you think?

Reply:
yup